### PR TITLE
Initial cookie design for webcontext

### PR DESCRIPTION
**Pull request by mmoayyed**
Closes https://github.com/pac4j/pac4j/issues/279

Let's review the current design, and I will continue to update pulls with docs, and tests.

### CODE REVIEW

**Comment by leleuj** [comment]
The build fails: `org.pac4j.http.credentials.CookieCredentials is not abstract and does not override abstract method hashCode() in org.pac4j.http.credentials.HttpCredentials`

File: pac4j-core/src/main/java/org/pac4j/core/context/WebContext.java [hunk 1]
```diff
@@ -162,9 +163,24 @@
...
-     * 
+     *
      * @return the URL
      * @since 1.5.0
      */
     String getFullRequestURL();
+
+    /**
+     * Retrieves request cookies.
+     * @since 1.8.1
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/context/WebContext.java]: If the PR is ready, let's do it for 1.8.0, no?

File: pac4j-http/src/main/java/org/pac4j/http/client/direct/CookieClient.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * @since 1.8.1
+ */
+public class CookieClient extends DirectHttpClient<CookieCredentials>  {
+    private String cookieName;
+    private String cookieValue;
+
+    public CookieClient() {
+    }
+
+    public CookieClient(final CookieAuthenticator cookieAuthenticator) {
```
Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/client/direct/CookieClient.java]: I think there should be a constructor with a cookie name.

File: pac4j-http/src/main/java/org/pac4j/http/client/direct/CookieClient.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import org.pac4j.http.credentials.extractor.CookieExtractor;
+import org.pac4j.http.profile.creator.ProfileCreator;
+
+/**
+ * @author Misagh Moayyed
+ * @since 1.8.1
+ */
+public class CookieClient extends DirectHttpClient<CookieCredentials>  {
+    private String cookieName;
+    private String cookieValue;
```
Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/client/direct/CookieClient.java]: I don't see why we need a `cookieValue`: it should be the returned value for the `CookieExtractor`, shouldn't it?

File: pac4j-http/src/main/java/org/pac4j/http/credentials/CookieCredentials.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+package org.pac4j.http.credentials;
+
+import org.pac4j.core.util.CommonHelper;
+
+/**
+ * @author Misagh Moayyed
+ * @since 1.8.1
+ */
+public class CookieCredentials extends HttpCredentials {
```
Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/credentials/CookieCredentials.java]: I would have expected to use the `TokenCredentials` set with the cookie value (search based on the name)

File: pac4j-http/src/main/java/org/pac4j/http/credentials/authenticator/CookieAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.http.credentials.CookieCredentials;
+
+/**
+ * @author Misagh Moayyed
+ * @since 1.8.1
+ */
+public class CookieAuthenticator implements Authenticator<CookieCredentials> {
```
Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/credentials/authenticator/CookieAuthenticator.java]: I don't think we need that by default: the cookie authenticator will be some dedicated implementation from the `TokenAuthenticator`.

File: pac4j-http/src/main/java/org/pac4j/http/credentials/extractor/CookieExtractor.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    private final String clientName;
+
+    public CookieExtractor(final String cookieName, final String cookieValue, final String clientName) {
+        this.cookieName = cookieName;
+        this.clientName = clientName;
+        this.cookieValue = cookieValue;
+    }
+
+    @Override
+    public CookieCredentials extract(final WebContext context) {
```
Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/credentials/extractor/CookieExtractor.java]: I would have expected this extractor to return the value of the found cookie (by its name).

File: pac4j-core/src/main/java/org/pac4j/core/context/WebContext.java (hunk 1 above)
Comment (mmoayyed) [inline pac4j-core/src/main/java/org/pac4j/core/context/WebContext.java]: Could be ready yes. If the schedule is by Thu, then we can make this work

**Comment by leleuj** [comment]
One more thing: do you think you can update the Play web context: https://github.com/pac4j/play-pac4j/blob/master/play-pac4j-java/src/main/java/org/pac4j/play/PlayWebContext.java? Or do you want me to handle that?

**Comment by mmoayyed** [comment]
I will review the comments here shortly, thanks for reviewing. As for play, go for it. I have not had much experience with play, and it would take me twice as long :)

**Comment by leleuj** [comment]
@mmoayyed : do you think you'll be able to finish that today (for the RC1 tomorrow) or do we keep it for 1.8.1?

**Comment by mmoayyed** [comment]
I think we should do this for 1.8.1. Sorry, caught in a very confusing troubleshooting session right now :( I wont have time until likely next monday.

**Comment by leleuj** [comment]
No problem.

**Comment by mmoayyed** [comment]
Might be able to resume working on this pull this afternoon. Let's check back tomorrow morning to see where I stand.

**Comment by mmoayyed** [comment]
OK, I am back to work on this. Stay tuned.

**Comment by mmoayyed** [comment]
Made a few updates. The problem I have now, is that we want authentication to act based on a cookie in the response, correct? This is not possible without drastic measures because the response does not expose a way to read cookies.

**Comment by mmoayyed** [comment]
Think I have that wrong. The authentication should work based on request cookies!

**Comment by mmoayyed** [comment]
Updated pull.

**Comment by leleuj** [comment]
Yes, request cookies. Will review.

**Comment by mmoayyed** [comment]
Not sure why the build fails! Everything compiles just fine for me. Maybe the build status is old?

**Comment by leleuj** [comment]
Regarding the build, you're not up-to-date with the master. I cleaned classes you are still referencing...

```
[ERROR] /home/travis/build/pac4j/pac4j/pac4j-http/src/test/java/org/pac4j/http/client/direct/CookieClientTests.java:[46,15] cannot find symbol
[ERROR] symbol:   class SimpleTestUsernameProfileCreator
```

File: pac4j-http/src/test/java/org/pac4j/http/client/direct/CookieClientTests.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ * @author Misagh Moayyed
+ * @since 1.8.0
+ */
+public class CookieClientTests implements TestsConstants {
+
+    @Test
+    public void testClone() {
+        final CookieClient oldClient = new CookieClient();
+        oldClient.setName(TYPE);
+        final SimpleTestUsernameProfileCreator profileCreator = new SimpleTestUsernameProfileCreator();
```
Comment (leleuj) [inline pac4j-http/src/test/java/org/pac4j/http/client/direct/CookieClientTests.java]: This profile creator does not exist anymore as the authenticator saves its profile in the credentials for the default `AuthenticatorProfileCreator`.

**Comment by leleuj** [comment]
+1 Just the test (and the build) must be fixed

**Comment by mmoayyed** [comment]
Merged.

**Comment by leleuj** [comment]
Build fixed. Will merge and update play-pac4j. Thanks.

**Pull request merged**